File: agenttel-go/processor/processor.go
// Package processor implements the OTel SpanProcessor that enriches spans with AgentTel metadata.
package processor

import (
	"context"
	"time"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"

	"go.agenttel.dev/agenttel/anomaly"
	"go.agenttel.dev/agenttel/attributes"
	"go.agenttel.dev/agenttel/baseline"
	"go.agenttel.dev/agenttel/errorclass"
	"go.agenttel.dev/agenttel/events"
	"go.agenttel.dev/agenttel/slo"
	"go.agenttel.dev/agenttel/topology"
)

// SpanCompletionListener is called after a span is processed.
type SpanCompletionListener func(span sdktrace.ReadOnlySpan)

// AgentTelSpanProcessor enriches OTel spans with AgentTel observability attributes.
type AgentTelSpanProcessor struct {
	baselineProvider baseline.Provider
	rollingProvider  *baseline.RollingProvider
	anomalyDetector  *anomaly.Detector
	patternMatcher   *anomaly.PatternMatcher
	sloTracker       *slo.Tracker
	errorClassifier  *errorclass.Classifier
	topologyRegistry *topology.Registry
	eventEmitter     *events.Emitter
	listener         SpanCompletionListener
}

// Option configures the span processor.
type Option func(*AgentTelSpanProcessor)

// WithBaseline sets the baseline provider for span enrichment.
func WithBaseline(p baseline.Provider) Option {
	return func(sp *AgentTelSpanProcessor) { sp.baselineProvider = p }
}

// WithRolling sets the rolling baseline provider.
func WithRolling(p *baseline.RollingProvider) Option {
	return func(sp *AgentTelSpanProcessor) { sp.rollingProvider = p }
}

// WithAnomalyDetector sets the anomaly detector.
func WithAnomalyDetector(d *anomaly.Detector) Option {
	return func(sp *AgentTelSpanProcessor) { sp.anomalyDetector = d }
}

// WithPatternMatcher sets the pattern matcher.
func WithPatternMatcher(m *anomaly.PatternMatcher) Option {
	return func(sp *AgentTelSpanProcessor) { sp.patternMatcher = m }
}

// WithSLOTracker sets the SLO tracker.
func WithSLOTracker(t *slo.Tracker) Option {
	return func(sp *AgentTelSpanProcessor) { sp.sloTracker = t }
}

// WithErrorClassifier sets the error classifier.
func WithErrorClassifier(c *errorclass.Classifier) Option {
	return func(sp *AgentTelSpanProcessor) { sp.errorClassifier = c }
}

// WithTopology sets the topology registry.
func WithTopology(r *topology.Registry) Option {
	return func(sp *AgentTelSpanProcessor) { sp.topologyRegistry = r }
}

// WithEventEmitter sets the event emitter.
func WithEventEmitter(e *events.Emitter) Option {
	return func(sp *AgentTelSpanProcessor) { sp.eventEmitter = e }
}

// WithCompletionListener sets a listener called after each span is processed.
func WithCompletionListener(l SpanCompletionListener) Option {
	return func(sp *AgentTelSpanProcessor) { sp.listener = l }
}

// New creates an AgentTelSpanProcessor with the given options.
func New(opts ...Option) *AgentTelSpanProcessor {
	p := &AgentTelSpanProcessor{}
	for _, opt := range opts {
		opt(p)
	}
	return p
}

// OnStart enriches spans with baseline and topology attributes at creation time.
func (p *AgentTelSpanProcessor) OnStart(_ context.Context, span sdktrace.ReadWriteSpan) {
	name := span.Name()

	// Enrich with topology
	if p.topologyRegistry != nil {
		if team := p.topologyRegistry.Team(); team != "" {
			span.SetAttributes(attribute.String(attributes.TopologyTeam, team))
		}
		if tier := p.topologyRegistry.Tier(); tier != "" {
			span.SetAttributes(attribute.String(attributes.TopologyTier, tier))
		}
		if domain := p.topologyRegistry.Domain(); domain != "" {
			span.SetAttributes(attribute.String(attributes.TopologyDomain, domain))
		}
	}

	// Enrich with baselines
	if p.baselineProvider != nil {
		if b, ok := p.baselineProvider.GetBaseline(name); ok {
			span.SetAttributes(
				attribute.Float64(attributes.BaselineLatencyP50Ms, b.LatencyP50Ms),
				attribute.Float64(attributes.BaselineLatencyP99Ms, b.LatencyP99Ms),
				attribute.Float64(attributes.BaselineErrorRate, b.ErrorRate),
				attribute.String(attributes.BaselineSource, b.Source),
				attribute.String(attributes.BaselineUpdatedAt, b.UpdatedAt.Format(time.RFC3339)),
			)
		}
	}
}

// OnEnd records rolling baselines, performs anomaly detection, and tracks SLOs.
func (p *AgentTelSpanProcessor) OnEnd(span sdktrace.ReadOnlySpan) {
	name := span.Name()
	durationMs := float64(span.EndTime().Sub(span.StartTime()).Milliseconds())
	isError := span.Status().Code == 2 // otel codes.Error = 2

	// Record rolling baseline
	if p.rollingProvider != nil {
		if isError {
			p.rollingProvider.RecordError(name)
		} else {
			p.rollingProvider.RecordLatency(name, durationMs)
		}
	}

	// SLO tracking
	if p.sloTracker != nil {
		if isError {
			p.sloTracker.RecordFailure(name)
		} else {
			p.sloTracker.RecordSuccess(name)
		}
	}

	// Anomaly detection
	if p.anomalyDetector != nil && p.rollingProvider != nil {
		if snap, ok := p.rollingProvider.GetSnapshot(name); ok && !snap.IsEmpty() {
			result := p.anomalyDetector.Evaluate("latency", durationMs, snap.Mean, snap.Stddev)
			if result.IsAnomaly && p.eventEmitter != nil {
				p.eventEmitter.Emit(attributes.EventAnomalyDetected, map[string]any{
					"operation":    name,
					"anomalyScore": result.AnomalyScore,
					"zScore":       result.ZScore,
					"latencyMs":    durationMs,
				})
			}
		}
	}

	// Notify listener
	if p.listener != nil {
		p.listener(span)
	}
}

// Shutdown gracefully shuts down the processor.
func (p *AgentTelSpanProcessor) Shutdown(_ context.Context) error {
	return nil
}

// ForceFlush flushes pending data.
func (p *AgentTelSpanProcessor) ForceFlush(_ context.Context) error {
	return nil
}

File: agenttel-node/src/processor.ts
import type { ReadableSpan, SpanProcessor } from '@opentelemetry/sdk-trace-node';
import type { Context } from '@opentelemetry/api';
import type { Span } from '@opentelemetry/api';
import * as attrs from './attributes.js';
import { AnomalyDetector } from './anomaly/detector.js';
import type { BaselineProvider } from './baseline/provider.js';
import { RollingBaselineProvider } from './baseline/rolling.js';
import { ErrorClassifier } from './error/classifier.js';
import { SLOTracker } from './slo/tracker.js';
import { TopologyRegistry } from './topology/registry.js';
import { AgentTelEventEmitter } from './events/emitter.js';

export interface ProcessorOptions {
  topology?: TopologyRegistry;
  baselineProvider?: BaselineProvider;
  rollingProvider?: RollingBaselineProvider;
  anomalyDetector?: AnomalyDetector;
  sloTracker?: SLOTracker;
  errorClassifier?: ErrorClassifier;
  eventEmitter?: AgentTelEventEmitter;
}

/** OTel SpanProcessor that enriches spans with AgentTel metadata. */
export class AgentTelSpanProcessor implements SpanProcessor {
  private readonly topology?: TopologyRegistry;
  private readonly baselineProvider?: BaselineProvider;
  private readonly rollingProvider?: RollingBaselineProvider;
  private readonly anomalyDetector?: AnomalyDetector;
  private readonly sloTracker?: SLOTracker;
  private readonly errorClassifier?: ErrorClassifier;
  private readonly eventEmitter?: AgentTelEventEmitter;

  constructor(options: ProcessorOptions = {}) {
    this.topology = options.topology;
    this.baselineProvider = options.baselineProvider;
    this.rollingProvider = options.rollingProvider;
    this.anomalyDetector = options.anomalyDetector;
    this.sloTracker = options.sloTracker;
    this.errorClassifier = options.errorClassifier;
    this.eventEmitter = options.eventEmitter;
  }

  onStart(span: Span, _parentContext: Context): void {
    // Enrich with topology metadata
    if (this.topology) {
      if (this.topology.team) span.setAttribute(attrs.TOPOLOGY_TEAM, this.topology.team);
      if (this.topology.tier) span.setAttribute(attrs.TOPOLOGY_TIER, this.topology.tier);
      if (this.topology.domain) span.setAttribute(attrs.TOPOLOGY_DOMAIN, this.topology.domain);

      const deps = this.topology.dependencies();
      if (deps.length > 0) {
        span.setAttribute(attrs.TOPOLOGY_DEPENDENCIES, this.topology.serializeDependenciesJSON());
      }
      const consumers = this.topology.consumers();
      if (consumers.length > 0) {
        span.setAttribute(attrs.TOPOLOGY_CONSUMERS, this.topology.serializeConsumersJSON());
      }
    }

    // Enrich with baseline data
    const opName = (span as unknown as { name: string }).name;
    if (this.baselineProvider && opName) {
      const baseline = this.baselineProvider.getBaseline(opName);
      if (baseline) {
        span.setAttribute(attrs.BASELINE_LATENCY_P50_MS, baseline.latencyP50Ms);
        span.setAttribute(attrs.BASELINE_LATENCY_P99_MS, baseline.latencyP99Ms);
        span.setAttribute(attrs.BASELINE_ERROR_RATE, baseline.errorRate);
        span.setAttribute(attrs.BASELINE_SOURCE, baseline.source);
      }
    }
  }

  onEnd(span: ReadableSpan): void {
    const opName = span.name;
    const durationMs = (span.endTime[0] - span.startTime[0]) * 1000 +
      (span.endTime[1] - span.startTime[1]) / 1e6;
    const isError = span.status?.code === 2; // StatusCode.ERROR

    // Record in rolling baseline
    if (this.rollingProvider && opName) {
      if (isError) {
        this.rollingProvider.recordError(opName);
      } else {
        this.rollingProvider.recordLatency(opName, durationMs);
      }
    }

    // Anomaly detection
    if (this.anomalyDetector && this.rollingProvider && opName) {
      const snap = this.rollingProvider.getSnapshot(opName);
      if (snap && snap.sampleCount > 0) {
        const result = this.anomalyDetector.evaluate('latency', durationMs, snap.mean, snap.stddev);
        if (result.isAnomaly && this.eventEmitter) {
          this.eventEmitter.emit(attrs.EVENT_ANOMALY_DETECTED, {
            operation: opName,
            score: result.anomalyScore,
            zScore: result.zScore,
          });
        }
      }
    }

    // SLO tracking
    if (this.sloTracker && opName) {
      if (isError) {
        this.sloTracker.recordFailure(opName);
      } else {
        this.sloTracker.recordSuccess(opName);
      }
    }

    // Error classification
    if (this.errorClassifier && isError) {
      const events = span.events || [];
      for (const event of events) {
        if (event.name === 'exception') {
          const exType = String(event.attributes?.['exception.type'] ?? '');
          const exMsg = String(event.attributes?.['exception.message'] ?? '');
          if (exType) {
            this.errorClassifier.classifyException(exType, exMsg);
          }
        }
      }
    }
  }

  async shutdown(): Promise<void> {
    // no-op
  }

  async forceFlush(): Promise<void> {
    // no-op
  }
}
